projects/rocprofiler-systems/source/lib/core/trace_cache/rocpd_post_processing.cpp

```diff
@@ -375,6 +376,219 @@ rocpd_post_processing::get_pmc_event_with_sample_callback() const
+                                trait::name<category::amd_smi_gfx_busy>::value,
+                                _amd_smi.gfx_activity);
+        insert_event_and_sample(is_busy_enabled,
+                                trait::name<category::amd_smi_umc_busy>::value,
+                                _amd_smi.umc_activity);
+        insert_event_and_sample(is_busy_enabled,
+                                trait::name<category::amd_smi_mm_busy>::value,
+                                _amd_smi.mm_activity);
+        insert_event_and_sample(is_temp_enabled,
+                                trait::name<category::amd_smi_temp>::value,
+                                _amd_smi.temperature);
+
+        insert_event_and_sample(is_power_enabled,
+                                trait::name<category::amd_smi_power>::value,
+                                _amd_smi.power);
+        insert_event_and_sample(is_mem_usage_enabled,
+                                trait::name<category::amd_smi_memory_usage>::value,
+                                _amd_smi.mem_usage);
+
+        if(is_vcn_enabled || is_jpeg_enabled)
+        {
+            xcp_metrics_t xcp_metrics;
+            deserialize_xcp_metrics(_amd_smi.xcp_activity, xcp_metrics);
+
+            if(is_vcn_enabled)
+            {
+                for(size_t clk = 0; clk < xcp_metrics.vcn_busy.size(); ++clk)
+                {
+                    const auto value = xcp_metrics.vcn_busy[clk];
+                    if(value == std::numeric_limits<uint16_t>::max())
+                    {
+                        continue;
+                    }
+
+                    std::stringstream ss;
+                    ss << trait::name<category::amd_smi_vcn_activity>::value;
+                    ss << "_" << std::to_string(clk);
+
+                    insert_event_and_sample(is_vcn_enabled, ss.str().c_str(), value);
+                }
+            }
+
+            if(is_jpeg_enabled)
+            {
+                for(size_t clk = 0; clk < xcp_metrics.jpeg_busy.size(); ++clk)
+                {
+                    const auto value = xcp_metrics.jpeg_busy[clk];
+                    if(value == std::numeric_limits<uint16_t>::max())
+                    {
+                        continue;
+                    }
+
+                    std::stringstream ss;
+                    ss << trait::name<category::amd_smi_jpeg_activity>::value;
+                    ss << "_" << std::to_string(clk);
+
+                    insert_event_and_sample(is_jpeg_enabled, ss.str().c_str(), value);
+                }
+            }
+        }
+    };
+}
+postprocessing_callback
+rocpd_post_processing::get_cpu_freq_sample_callback() const
+{
+    struct core_freq_sample
+    {
+        size_t id;
+        float  value;
+    };
+
+    auto deserialize_freqs = [](std::vector<uint8_t>& data) {
+        std::vector<core_freq_sample> result;
+        size_t                        offset = 0;
+
+        while(offset != data.size())
```

Comment:
```suggestion
        while(offset < data.size())
```
May be safer.
Or, even safer:

```suggestion
        while(offset + sizeof(float) + sizeof(size_t) <= data.size())
```